src/main/scala/akka/stream/contrib/DirectoryChanges.scala

```diff
@@ -0,0 +1,162 @@
+  case object Modified extends Change
+  case object Created extends Change
+  case object Deleted extends Change
+
+  /**
+   * Java API
+   */
+  val modified = Modified
+  /**
+   * Java API
+   */
+  val created = Created
+  /**
+   * Java API
+   */
+  val deleted = Deleted
+
+  /**
+   * Scala API
+   *
+   * Watches a file system directory and streams change events from it.
+   *
+   * @see [[DirectoryChanges]] for details
+   */
+  def apply(directory: Path, pollInterval: FiniteDuration = 500.millis, maxBufferSize: Int = 200): Source[(Path, Change), NotUsed] =
+    Source.fromGraph(new DirectoryChanges(directory, pollInterval, maxBufferSize))
+
+  /**
+   * Java API
+   * @see [[DirectoryChanges]] for details
+   */
+  def forDirectory(directory: Path): Source[Pair[Path, DirectoryChanges.Change], NotUsed] = apply(directory).map(t => Pair.create(t._1, t._2))
+
+  /**
+   * Java API
+   * @see [[DirectoryChanges]] for details
+   */
+  def forDirectory(directory: Path, pollInterval: FiniteDuration, maxBufferSize: Int): Source[Pair[Path, DirectoryChanges.Change], NotUsed] =
+    apply(directory, pollInterval, maxBufferSize).map(t => Pair.create(t._1, t._2))
+
+  private def japiKindToChange(orig: WatchEvent.Kind[Path]): Change = orig match {
+    case StandardWatchEventKinds.ENTRY_CREATE => Created
+    case StandardWatchEventKinds.ENTRY_DELETE => Deleted
+    case StandardWatchEventKinds.ENTRY_MODIFY => Modified
+    case _                                    => throw new IllegalArgumentException(s"unknown event kind $orig")
+  }
+
+  private val defaultAttributes = Attributes.name("DirectoryChanges")
+
+}
+
+/**
+ * Watches a file system directory and streams change events from it.
+ *
+ * Note that the JDK watcher is notoriously slow on some platform (up to 1s after event actually happened on OSX for example)
+ *
+ * @param directoryPath Directory to watch
+ * @param pollInterval Interval between polls to the JDK watch service when a push comes in and there was no changes, if
+ *                     the JDK implementation is slow, it will not help lowering this
+ * @param maxBufferSize Maximum number of buffered directory changes before the stage fails
+ */
+final class DirectoryChanges(directoryPath: Path, pollInterval: FiniteDuration, maxBufferSize: Int) extends GraphStage[SourceShape[(Path, DirectoryChanges.Change)]] {
+  import DirectoryChanges._
+  private val out: Outlet[(Path, Change)] = Outlet("DirectoryChanges.out")
+
+  override protected def initialAttributes: Attributes = defaultAttributes
+
+  override def shape: SourceShape[(Path, Change)] = SourceShape.of(out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
+    assert(Files.exists(directoryPath), s"The path: $directoryPath does not exist")
+    assert(Files.isDirectory(directoryPath), s"The path $directoryPath is not a directory")
+
+    private var buffer = Queue.empty[(Path, Change)]
```

Comment:
Yes, performance. As for the strategy, I don't even think any other strategy than fail makes sense here, since dropped events result in out-of sync view. A downstream fused dropping buffer can be added by the user if he for some reason still prefers dropping.